plugins/metrics/pkg/agent/runner.go

```diff
@@ -203,96 +195,157 @@ func (tr *taskRunner) doPush(ctx context.Context, writeRequest *prompb.WriteRequ
+	waitctx.Go(wc, func() {
+		labelMatchers := toLabelMatchers(run.Query.Matchers)
 
-	progressDelta := nextStart
+		importEnd := run.Query.EndTimestamp.AsTime().UnixMilli()
+		nextStart := run.Query.StartTimestamp.AsTime().UnixMilli()
+		nextEnd := nextStart
 
-	progress := &corev1.Progress{
-		Current: 0,
-		Total:   uint64(importEnd - progressDelta),
-	}
+		reader := NewRemoteReader(&http.Client{})
 
-	activeTask.SetProgress(progress)
+		for err == nil && nextStart < importEnd {
+			select {
+			case <-wc.Done():
+				return
+			default: // continue reading
+			}
 
-	activeTask.AddLogEntry(zapcore.InfoLevel, "import running")
+			nextStart = nextEnd
+			nextEnd = nextStart + TimeDeltaMillis
 
-	for nextStart < importEnd {
-		select {
-		case <-ctx.Done():
-			activeTask.AddLogEntry(zapcore.InfoLevel, "import stopped")
-			return ctx.Err()
-		default: // continue with import
-		}
+			if nextStart >= importEnd {
+				break
+			}
 
-		nextStart = nextEnd
-		nextEnd = nextStart + TimeDeltaMillis
+			if nextEnd >= importEnd {
+				nextEnd = importEnd
+			}
 
-		if nextStart >= importEnd {
-			break
-		}
+			readRequest := &prompb.ReadRequest{
+				Queries: []*prompb.Query{
+					{
+						StartTimestampMs: nextStart,
+						EndTimestampMs:   nextEnd,
+						Matchers:         labelMatchers,
+					},
+				},
+			}
 
-		if nextEnd >= importEnd {
-			nextEnd = importEnd
-		}
+			var readResponse *prompb.ReadResponse
 
-		readRequest := &prompb.ReadRequest{
-			Queries: []*prompb.Query{
-				{
-					StartTimestampMs: nextStart,
-					EndTimestampMs:   nextEnd,
-					Matchers:         labelMatchers,
-				},
-			},
-		}
+			readResponse, err = tr.doRead(wc, reader, run, readRequest)
+			if err != nil {
+				return
+			}
 
-		readResponse, err := tr.remoteReader.Read(context.Background(), run.Target.Spec.Endpoint, readRequest)
+			writeRequest := &prompb.WriteRequest{
+				Timeseries: lo.Map(readResponse.Results[0].GetTimeseries(), func(t *prompb.TimeSeries, _ int) prompb.TimeSeries {
+					return lo.FromPtr(t)
+				}),
+				Metadata: []prompb.MetricMetadata{},
+			}
 
-		if err != nil {
-			activeTask.AddLogEntry(zapcore.ErrorLevel, fmt.Sprintf("failed to read from target endpoint: %s", err))
-			return fmt.Errorf("failed to read from target endpoint: %w", err)
-		}
+			var chunks []*prompb.WriteRequest
 
-		for _, result := range readResponse.Results {
-			if len(result.Timeseries) == 0 {
-				continue
+			chunks, err = util.SplitChunksWithLimit(writeRequest, limit)
+			if err != nil {
+				return
 			}
 
-			writeRequest := prompb.WriteRequest{
-				Timeseries: dereferenceResultTimeseries(result.Timeseries),
+			activeTask.AddLogEntry(zapcore.InfoLevel, fmt.Sprintf("split request into %d chunks", len(chunks)))
+
+			lo.ForEach(chunks, func(chunk *prompb.WriteRequest, i int) {
+				if err := b.Add(wc, WriteMetadata{
+					Query:         readRequest.Queries[0],
+					WriteChunk:    chunk,
+					ProgressRatio: 1.0 / float64(len(chunks)),
```

Comment:
Do all chunks have the same progress ratio?

Reply:
All chunks from the same original `prompb.WriteRequest` do, but will differ between chunks from different `prompb.WriteRequest`.

Not super happy with the way I currently have this implemented. Will probably change this to a single progress delta so I can get rid of the `Query` field and the `ProgressRatio` all the same logic will remain essentially the same just moved to before data is added to the buffer, but we'll be storing less data.